src/Microsoft.Diagnostics.EventFlow.Outputs.EventHub/EventHubOutput.cs

```diff
@@ -83,52 +84,67 @@ public EventHubOutput(IConfiguration configuration, IHealthReporter healthReport
 
             try
             {
-                // Since event hub limits each message/batch to be a certain size, we need to
-                // keep checking the size for exceeds and split into a new batch as needed
+                var groupedEventData = events.GroupBy(e => string.IsNullOrEmpty(outputConfiguration.PartitionKeyProperty) == false && e.TryGetPropertyValue(outputConfiguration.PartitionKeyProperty, out var partionKeyData) ? partionKeyData.ToString() : string.Empty, e => e);
 
-                List<List<MessagingEventData>> batches = new List<List<MessagingEventData>>();
-                int batchByteSize = 0;
+                List<Task> tasks = new List<Task>();
 
-                foreach (EventData eventData in events)
+                foreach (var partitionedEventData in groupedEventData)
                 {
-                    int messageSize;
-                    MessagingEventData messagingEventData = eventData.ToMessagingEventData(out messageSize);
+                    // Since event hub limits each message/batch to be a certain size, we need to
+                    // keep checking the size for exceeds and split into a new batch as needed
+
+                    List<List<MessagingEventData>> batches = new List<List<MessagingEventData>>();
+                    int batchByteSize = 0;
 
-                    // If we don't have a batch yet, or the addition of this message will exceed the limit for this batch, then
-                    // start a new batch.
-                    if (batches.Count == 0 ||
-                        batchByteSize + messageSize > EventHubMessageSizeLimit)
+                    foreach (EventData eventData in partitionedEventData)
                     {
-                        batches.Add(new List<MessagingEventData>());
-                        batchByteSize = 0;
+                        int messageSize;
+                        MessagingEventData messagingEventData = eventData.ToMessagingEventData(out messageSize);
+
+                        // If we don't have a batch yet, or the addition of this message will exceed the limit for this batch, then
+                        // start a new batch.
+                        if (batches.Count == 0 ||
+                            batchByteSize + messageSize > EventHubMessageSizeLimit)
+                        {
+                            batches.Add(new List<MessagingEventData>());
+                            batchByteSize = 0;
+                        }
+
+                        batchByteSize += messageSize;
+
+                        List<MessagingEventData> currentBatch = batches[batches.Count - 1];
```

Comment:
This is pre-existing issue, but can you move this assignment into the if statement body above, i.e. after line 110? We do not need to do this for every EventData instance, only when we create a new batch